cord/user_client.py

```diff
@@ -75,6 +85,221 @@ def create_project_api_key(self, project_hash: str, api_key_title: str, scopes:
+            return CvatImporterSuccess(
+                project_hash=success["project_hash"],
+                dataset_hash=dataset_hash,
+                issues=Issues.from_dict(success["issues"])
+            )
+        elif "error" in server_ret:
+            error = server_ret["error"]
+            return CvatImporterError(
+                dataset_hash=dataset_hash,
+                issues=Issues.from_dict(error["issues"])
+            )
+        else:
+            raise ValueError("The api server responded with an invalid payload.")
+
+    def __get_images_paths(self, annotations_base64: str, images_directory_path: Path) -> List[Path]:
+        payload = {
+            "annotations_base64": annotations_base64
+        }
+        project_info = self.querier.basic_setter(ProjectImporterCvatInfo, uid=None, payload=payload)
+        if "error" in project_info:
+            message = project_info["error"]["message"]
+            raise ValueError(message)
+
+        export_type = project_info["success"]["export_type"]
+        if export_type == CvatExportType.PROJECT.value:
+            default_path = images_directory_path.joinpath('default')
+            if default_path not in list(images_directory_path.iterdir()):
+                raise ValueError("The expected directory 'default' was not found.")
+
+            images = list(default_path.iterdir())
+        elif export_type == CvatExportType.TASK.value:
+            images = list(images_directory_path.iterdir())
+        else:
+            raise ValueError("Received an unexpected response from the server. Project import aborted.")
+
+        if not images:
+            raise ValueError(f"No images found in the provided data folder.")
+        return images
+
+    def __upload_cvat_images(self,
+                             images_paths: List[Path],
+                             dataset_name: str) -> Tuple[str, Dict[str, str]]:
+        """
+        This function does not create any image groups yet.
+        Returns:
+            * The created dataset_hash
+            * A map from an image title to the image hash which is stored in the DB.
+        """
+
+        short_names = list(map(lambda x: x.name, images_paths))
+        file_path_strings = list(map(lambda x: str(x), images_paths))
+        dataset = self.create_dataset(dataset_name, DatasetType.CORD_STORAGE)
+
+        dataset_hash = dataset["dataset_hash"]
+
+        dataset_api_key: DatasetAPIKey = self.create_dataset_api_key(
+            dataset_hash,
+            dataset_name + " - Full Access API Key",
+            [DatasetScope.READ, DatasetScope.WRITE])
+
+        client = CordClient.initialise(
+            dataset_hash,
+            dataset_api_key.api_key,
+            domain=self.user_config.domain,
+        )
+
+        signed_urls = client._querier.basic_getter(
+            SignedImagesURL,
+            uid=short_names
+        )
+        asyncio.run(upload_to_signed_url_list_async(
```

Comment:
What's the reasoning for blocking on an async call? Seems a bit ... unnecessary?

Reply:
You cannot await a coroutine (i.e. async function) from within a normal (non-async) function. However, because we're calling `upload_to_signed_url_list_async` with `asyncio.run`, we can then within `upload_to_signed_url_list_async` do all sorts of things we can usually do with coroutines, like uploading all images in batch.

The way I interpret it is "allow uploading in parallel, but don't continue from this line until all uploads are done"